libs/mve/depthmap.h

```diff
@@ -167,16 +167,17 @@ MVE_IMAGE_NAMESPACE_BEGIN
 template <typename T>
 inline void
 depthmap_convert_conventions (typename Image<T>::Ptr dm,
-    math::Matrix3f const& invproj, bool to_mve)
+    math::Matrix3f const& invproj, bool to_mve, bool add_pixel_offset)
```

Comment:
I'm curious, why did you add this? This pixel offset is always necessary when converting from image coordinates (where the center of the top left pixel is (0,0)) to 3D image plane coordinates (where the center of the top left pixel is (0.5,0.5)).

Reply:
From the following comment in [colmap/blob/dev/src/base/reconstruction.cc](https://github.com/colmap/colmap/blob/dev/src/base/reconstruction.cc#L1036),

      // Bundler output assumes image coordinate system origin
      // in the lower left corner of the image with the center of
      // the lower left pixel being (0, 0). Our coordinate system
      // starts in the upper left corner with the center of the
      // upper left pixel being (0.5, 0.5).

I understand that the **image** coordinate system of Colmap also starts at (0.5,0.5). I made some experiments a few weeks ago, which "confirmed" that.

Reply:
Quick question: Does MVE offer some way to associate something like view_0006.mve with the original image name? Since the images are also reordered it is not easy to match the results with the original images

Reply:
Why is it relevant what convention Colmap uses internally for images? You're inputting images into MVE, and the pixels of that depth image correspond to the pixels of the color image. Now you're treating these images with MVE's convention to convert to a depth map in MVE's convention.

Unless you're telling me that that the depth map is shifted by 0.5 pixels compared to the color image, no change should be necessary here.

Each MVE view has a `meta.ini` file that has a `name` key. Typically you put the original image name (such as `IMG_123.JPG`) into that key. Does that help?

Reply:
The Colmap Fork I'm working with provides some matrices, which transform the reparameterized depth maps back into world coordinates. In the following, I'll use the corresponding point clouds as reference (shown in red).

Here, an overlay of the reference point cloud and the point cloud generated with offset=0.0 (green). The reason it does not completely match is the fact that the computations in the reparameterized space are numerically more stable. But one can see that the differences (the noise) appears to be random and non-systematic.

![no_offset](https://user-images.githubusercontent.com/19491335/92643422-09d97b80-f2e2-11ea-9417-834fe35d075d.jpg)

Here, an overlay of the reference point cloud and the point cloud generated with offset=0.5 (green)
![with_offset](https://user-images.githubusercontent.com/19491335/92643606-558c2500-f2e2-11ea-9e2b-197c2bdfa0d2.jpg)
 
On can clearly see that there is a systematic difference between the point clouds.

I think that makes sense, or am I missing something here?

Reply:
But hold on a second. The offset we're talking about here is solely used to compute the scaling factor for converting depth values. I would assume that this offset has a negligible effect on the depth actually.

If the effect was actually visible, wouldn't we see slightly smaller depth values on the left side of the image (in front of geometry), and slightly larger depth values on the right side of the image (behind geometry)? So wouldn't a systematic artifact manifest in a greenish image on the left half, and a reddish image on the right half?

I'm not sure that's the problem we're seeing here.

Then again, there is also a chance that the Colmap generated geometry is shifted when raising the points to 3D space?

Reply:
Suggestion. Let's submit your PR without this boolean first. Then let's discuss this issue in a separate PR.

Reply:
Do you maybe have to add/subtract 0.5 from the principal point (before normalizing it) when Colmaps image conventions are different?

Reply:
> Why is it relevant what convention Colmap uses internally for images? You're inputting images into MVE, and the pixels of that depth image correspond to the pixels of the color image. Now you're treating these images with MVE's convention to convert to a depth map in MVE's convention. Unless you're telling me that Colmap's depth map is shifted by 0.5 pixels compared to the color image, no change should be necessary here.

I understand it as follows:
Obviously, each value in a depth map (together with the camera parameters) represents a 3D point in camera / world coordinates. Thus,  the meaning of the actual value is completely depending on the the other parameters. Assuming that colmap defined the depth values with respect to 
`math::Vec3f px((float)x, (float)y, 1.0f); ` instead of `math::Vec3f px((float)x + 0.5, (float)y + 0.5, 1.0f);`, I need to use exactly this computation to derive the original point in camera/world coordinates. 

Because of the comment in `reconstruction.cc`, I expected (assuming that Colmap images and Colmap depth maps are aligned) that `math::Vec3f px((float)x, (float)y, 1.0f); ` is the correct computation

Reply:
Let me clarify what convention MVE uses for images, in case I didn't make that clear.

When accessing image (such as mve::ByteImage), `(0,0)` is the center of the first pixel. And when you use linear interpolation, for example, `(0,0)` will give you the first pixel's color (and not an interpolated color). `(0.5, 0)` will give you the interpolation between the first and second pixel. The horizontal index starts at `0` and ends at `width - 1`. This implies that the valid image region is in range `[-0.5, width - 0.5] x [-0.5, height - 0.5]`. I think this is pretty standard convention.

Now, when we're talking about the 3D space, and looking at the image plane in 3D space, that plane has dimensions `[0, width] x [0, height]`, and the upper left corner corresponds to `(-0.5, -0.5)` in image coordinates. And when you convert pixel `(x,y)` to a 3D point you'll get a coordinate `(x + 0.5, y + 0.5, flen)`.

I hope this makes sense.

Reply:
> But hold on a second. The offset we're talking about here is solely used to compute the scaling factor for converting depth values. I would assume that this offset has a negligible effect on the depth actually.

Why? With this offset the position of the 3D point changes. And MVE uses the distance to the point as depth, so the value will be different. No?

> If the effect was actually visible, wouldn't we see slightly smaller depth values on the left side of the image (in front of geometry), and slightly larger depth values on the right side of the image (behind geometry)? So wouldn't a systematic artifact manifest in a greenish image on the left half, and a reddish image on the right half?

Acutally, I don't dare to make any speculations about this, since the camera models are so crazy. See this intrinsic parameters as example: `fx, fy, skew, cx, cy: [4150290.730193433, 4082933.0617155796, -5788.048198890227, 46942.580589824625, 718815.7955793234]`
Note, that I remove the skew before processing the results with MVE.

> Suggestion. Let's submit your PR without this boolean first. 

Okay

Reply:
> Then let's discuss this issue in a separate PR.

Actually, I have no idea how we could further verify this. So what would you expect to be contained in this separate PR?
Since you are the maintainer of the repository, I think you should decide how we handle this ;)

Reply:
> >     But hold on a second. The offset we're talking about here is solely used to compute the scaling factor for converting depth values. I would assume that this offset has a negligible effect on the depth actually.
> 
> Why? With this offset the position of the 3D point changes. And MVE uses the distance to the point as depth, so the value will be different. No?

This referred to the scaling value

Reply:
In this code

```
  math::Vec3f px((float)x + 0.5f, (float)y + 0.5f, 1.0f);
  px = invproj * px;
  double len = px.norm();
  dm->at(pos) *= (to_mve ? len : 1.0 / len);
```

the position of `px` does change with the offset, but it's really only used to scale the depth value (`px.norm()`), i.e., changes the final point cloud position in viewing direction, not orthogonal to it. Have you tried adding/subtracting 0.5 from the principal point before normalizing it (as per my previous comment). Maybe this fixes this issue you're seeing?

Reply:
This is to say, if the `0.5` offset was the problem, there would be more places you'd have to fix.
Most notably [this one](https://github.com/simonfuhrmann/mve/blob/master/libs/mve/depthmap.cc#L150).

Reply:
> Have you tried adding/subtracting 0.5 from the principal point before normalizing it (as per my previous comment). Maybe this fixes this issue you're seeing

No, but this could / should have the same effect. 

> This is to say, if the 0.5 offset was the problem, there would be more places you'd have to fix.

I'm afraid that's the case. 

Oh boy this importer is killing me ... :D

Reply:
I feel you man. These two things are among the hardest in human history: Converting between conventions, and rotations. Literally the hardest thing in the world is converting rotation conventions.